src/Core/BackgroundSchedulePool.cpp

```diff
@@ -361,15 +389,45 @@ void BackgroundSchedulePool::threadFunction()
             if (shutdown)
                 break;
 
-            if (!tasks.empty())
+            if (runnable_task_types.empty())
+                continue;
+
+            task_type_to_run = runnable_task_types[thread_local_rng() % runnable_task_types.size()];
```

Comment:
A round-robin over groups might be a bit more fair and faster because we are under lock here anyway